File: frontend/src/tableSync.ts
import type { Row, Tables } from './types';

const SLOT_ANNOTATION_COLUMN_PREFIX = 'Annotation: ';
const LEGACY_SLOT_ANNOTATION_COLUMN_PREFIX = 'annotation_';

const LEGACY_SLOT_ANNOTATION_COLUMNS: Record<string, string> = {
  mimicc_default_unit: 'default_unit'
};

const LEGACY_SLOT_ROW_COLUMNS: Record<string, string> = {
  annotation_mimicc_default_unit: 'Annotation: default_unit'
};

type SyncOptions = {
  sourceTable?: string;
};

export function syncTables(previousTables: Tables, nextTables: Tables, options: SyncOptions = {}): Tables {
  const tables = cloneTables(nextTables);
  for (const name of ['classes', 'slots', 'enums', 'permissible_values', 'annotations']) {
    tables[name] ??= [];
  }

  const renamed = syncRenames(previousTables, tables, options.sourceTable);
  syncDeletedReferences(previousTables, tables, renamed, options.sourceTable);
  pruneMissingReferences(tables);
  syncClassSlotMembership(tables, options.sourceTable);
  syncSlotAnnotations(tables, options.sourceTable);
  syncEnumAnnotations(tables, options.sourceTable);
  pruneMissingReferences(tables);
  return tables;
}

function syncRenames(previousTables: Tables, tables: Tables, sourceTable?: string) {
  const renamed = {
    classes: new Set<string>(),
    slots: new Set<string>(),
    enums: new Set<string>()
  };
  if (sourceTable === 'classes') {
    forEachRowRename(previousTables.classes, tables.classes, 'class', (oldName, nextName) => {
      renamed.classes.add(oldName);
      replaceCellValue(tables.slots, 'class', oldName, nextName);
      replaceAnnotationTarget(tables.annotations, 'class', oldName, nextName);
    });
  }

  if (sourceTable === 'slots') {
    forEachRowRename(previousTables.slots, tables.slots, 'slot', (oldName, nextName) => {
      renamed.slots.add(oldName);
      replaceAnnotationTarget(tables.annotations, 'slot', oldName, nextName);
    });
  }

  if (sourceTable === 'enums') {
    forEachRowRename(previousTables.enums, tables.enums, 'enum', (oldName, nextName) => {
      renamed.enums.add(oldName);
      replaceCellValue(tables.permissible_values, 'enum', oldName, nextName);
      replaceCellValue(tables.slots, 'range', oldName, nextName);
      replaceAnnotationTarget(tables.annotations, 'enum', oldName, nextName);
    });
  }
  return renamed;
}

function syncClassSlotMembership(tables: Tables, sourceTable?: string) {
  if (sourceTable === 'classes') {
    for (const classRow of tables.classes ?? []) {
      const className = cellText(classRow.class);
      const slotNames = splitList(classRow.slots);
      if (!className) continue;
      tables.slots = rebuildClassSlotRows(tables.slots ?? [], className, slotNames);
    }
    return;
  }

  const slotsByClass = new Map<string, string[]>();
  for (const row of tables.slots ?? []) {
    const className = cellText(row.class);
    const slotName = cellText(row.slot);
    if (className && slotName) {
      slotsByClass.set(className, [...(slotsByClass.get(className) ?? []), slotName]);
    }
  }
  tables.classes = (tables.classes ?? []).map((row) => {
    const className = cellText(row.class);
    return slotsByClass.has(className) ? { ...row, slots: slotsByClass.get(className)?.join('; ') ?? '' } : row;
  });
}

function syncSlotAnnotations(tables: Tables, sourceTable?: string) {
  migrateLegacySlotRowColumns(tables);
  migrateLegacySlotAnnotations(tables);

  if (sourceTable !== 'annotations') {
    for (const row of tables.slots ?? []) {
      const slotName = cellText(row.slot);
      if (!slotName) continue;
      for (const rowKey of slotAnnotationColumns(row)) {
        const annotationKey = annotationKeyForColumn(rowKey);
        upsertAnnotation(tables, 'slot', slotName, annotationKey, row[rowKey]);
      }
    }
  } else {
    for (const row of tables.slots ?? []) {
      for (const rowKey of slotAnnotationColumns(row)) {
        row[rowKey] = '';
      }
    }
  }

  const slotRows = new Map((tables.slots ?? []).map((row) => [cellText(row.slot), row]));
  for (const row of tables.annotations ?? []) {
    if (cellText(row.element_type) !== 'slot') continue;
    const slotRow = slotRows.get(cellText(row.element));
    const rowKey = annotationColumnForKey(cellText(row.key));
    if (slotRow && rowKey) {
      slotRow[rowKey] = row.value ?? '';
    }
  }
}

function syncEnumAnnotations(tables: Tables, sourceTable?: string) {
  if (sourceTable !== 'annotations') {
    for (const row of tables.enums ?? []) {
      const enumName = cellText(row.enum);
      if (!enumName) continue;
      const annotations = parseMapping(row.annotations);
      removeEnumAnnotationRowsNotIn(tables, enumName, new Set(Object.keys(annotations)));
      for (const [key, value] of Object.entries(annotations)) {
        upsertAnnotation(tables, 'enum', enumName, key, value);
      }
    }
  }

  const enumRows = new Map((tables.enums ?? []).map((row) => [cellText(row.enum), row]));
  for (const row of tables.annotations ?? []) {
    if (cellText(row.element_type) !== 'enum') continue;
    const enumRow = enumRows.get(cellText(row.element));
    const key = cellText(row.key);
    if (!enumRow || !key) continue;
    const annotations = parseMapping(enumRow.annotations);
    annotations[key] = row.value ?? '';
    enumRow.annotations = Object.keys(annotations).length ? JSON.stringify(annotations) : '';
  }
}

function syncDeletedReferences(
  previousTables: Tables,
  tables: Tables,
  renamed: ReturnType<typeof syncRenames>,
  sourceTable?: string
) {
  if (sourceTable === 'classes') {
    const deletedClasses = deletedNames(previousTables.classes, tables.classes, 'class', renamed.classes);
    if (deletedClasses.size) {
      tables.slots = (tables.slots ?? []).filter((row) => !deletedClasses.has(cellText(row.class)));
      tables.annotations = (tables.annotations ?? []).filter(
        (row) => !(cellText(row.element_type) === 'class' && deletedClasses.has(cellText(row.element)))
      );
    }
  }

  if (sourceTable === 'slots') {
    const deletedSlots = deletedNames(previousTables.slots, tables.slots, 'slot', renamed.slots);
    if (deletedSlots.size) {
      tables.annotations = (tables.annotations ?? []).filter(
        (row) => !(cellText(row.element_type) === 'slot' && deletedSlots.has(cellText(row.element)))
      );
    }
  }

  if (sourceTable === 'enums') {
    const deletedEnums = deletedNames(previousTables.enums, tables.enums, 'enum', renamed.enums);
    if (deletedEnums.size) {
      tables.permissible_values = (tables.permissible_values ?? []).filter(
        (row) => !deletedEnums.has(cellText(row.enum))
      );
      tables.annotations = (tables.annotations ?? []).filter(
        (row) => !(cellText(row.element_type) === 'enum' && deletedEnums.has(cellText(row.element)))
      );
      for (const row of tables.slots ?? []) {
        if (deletedEnums.has(cellText(row.range))) {
          row.range = '';
        }
      }
    }
  }
}

function pruneMissingReferences(tables: Tables) {
  const classNames = new Set((tables.classes ?? []).map((row) => cellText(row.class)).filter(Boolean));
  const enumNames = new Set((tables.enums ?? []).map((row) => cellText(row.enum)).filter(Boolean));

  tables.slots = (tables.slots ?? []).filter((row) => {
    const className = cellText(row.class);
    return !className || classNames.has(className);
  });
  for (const row of tables.slots ?? []) {
    const enumName = cellText(row.range);
    if (enumName.endsWith('Menu') && !enumNames.has(enumName)) {
      row.range = '';
    }
  }

  const slotNames = new Set((tables.slots ?? []).map((row) => cellText(row.slot)).filter(Boolean));
  tables.permissible_values = (tables.permissible_values ?? []).filter((row) => {
    const enumName = cellText(row.enum);
    return !enumName || enumNames.has(enumName);
  });
  tables.annotations = (tables.annotations ?? []).filter((row) =>
    annotationTargetExists(row, classNames, slotNames, enumNames)
  );
}

function annotationTargetExists(row: Row, classNames: Set<string>, slotNames: Set<string>, enumNames: Set<string>) {
  const elementType = cellText(row.element_type);
  const element = cellText(row.element);
  if (!elementType || !element) return true;
  if (elementType === 'class') return classNames.has(element);
  if (elementType === 'slot') return slotNames.has(element);
  if (elementType === 'enum') return enumNames.has(element);
  return true;
}

function removeEnumAnnotationRowsNotIn(tables: Tables, enumName: string, keys: Set<string>) {
  tables.annotations = (tables.annotations ?? []).filter(
    (row) =>
      !(
        cellText(row.element_type) === 'enum' &&
        cellText(row.element) === enumName &&
        !keys.has(cellText(row.key))
      )
  );
}

function rebuildClassSlotRows(slotRows: Row[], className: string, slotNames: string[]) {
  const unrelatedRows = slotRows.filter((row) => cellText(row.class) !== className);
  const currentClassRows = slotRows.filter((row) => cellText(row.class) === className);
  const allRowsBySlot = new Map(slotRows.map((row) => [cellText(row.slot), row]));
  const classRowsBySlot = new Map(currentClassRows.map((row) => [cellText(row.slot), row]));
  const orderedRows = slotNames.map((slotName, index) => ({
    ...(allRowsBySlot.get(slotName) ?? {}),
    ...(classRowsBySlot.get(slotName) ?? {}),
    class: className,
    slot: slotName,
    rank: cellText(classRowsBySlot.get(slotName)?.rank) || String(index + 1)
  }));
  return [...unrelatedRows, ...orderedRows];
}

function forEachRowRename(
  previousRows: Row[] | undefined,
  nextRows: Row[] | undefined,
  column: string,
  onRename: (oldName: string, nextName: string) => void
) {
  const previous = previousRows ?? [];
  const next = nextRows ?? [];
  if (previous.length !== next.length) return;
  for (let index = 0; index < Math.min(previous.length, next.length); index += 1) {
    const oldName = cellText(previous[index]?.[column]);
    const nextName = cellText(next[index]?.[column]);
    if (oldName && nextName && oldName !== nextName) {
      onRename(oldName, nextName);
    }
  }
}

function deletedNames(
  previousRows: Row[] | undefined,
  nextRows: Row[] | undefined,
  column: string,
  renamed: Set<string>
) {
  const nextNames = new Set((nextRows ?? []).map((row) => cellText(row[column])).filter(Boolean));
  return new Set(
    (previousRows ?? [])
      .map((row) => cellText(row[column]))
      .filter((name) => name && !renamed.has(name) && !nextNames.has(name))
  );
}

function replaceCellValue(rows: Row[] | undefined, column: string, oldValue: string, nextValue: string) {
  for (const row of rows ?? []) {
    if (cellText(row[column]) === oldValue) {
      row[column] = nextValue;
    }
  }
}

function replaceAnnotationTarget(
  rows: Row[] | undefined,
  elementType: string,
  oldElement: string,
  nextElement: string
) {
  for (const row of rows ?? []) {
    if (cellText(row.element_type) === elementType && cellText(row.element) === oldElement) {
      row.element = nextElement;
    }
  }
}

function upsertAnnotation(
  tables: Tables,
  elementType: string,
  element: string,
  key: string,
  value: Row[string]
) {
  const annotations = tables.annotations ?? [];
  const match = annotations.find(
    (row) =>
      cellText(row.element_type) === elementType &&
      cellText(row.element) === element &&
      cellText(row.key) === key
  );
  if (value === null || value === undefined || value === '') {
    if (match) {
      tables.annotations = annotations.filter((row) => row !== match);
    }
    return;
  }
  if (match) {
    match.value = value;
  } else {
    annotations.push({ element_type: elementType, element, key, value });
    tables.annotations = annotations;
  }
}

function migrateLegacySlotRowColumns(tables: Tables) {
  for (const row of tables.slots ?? []) {
    migrateLegacySlotAnnotationColumns(row);
    for (const [legacyKey, rowKey] of Object.entries(LEGACY_SLOT_ROW_COLUMNS)) {
      if (!cellText(row[rowKey]) && cellText(row[legacyKey])) {
        row[rowKey] = row[legacyKey];
      }
      delete row[legacyKey];
    }
  }
}

function migrateLegacySlotAnnotations(tables: Tables) {
  for (const row of tables.annotations ?? []) {
    if (cellText(row.element_type) !== 'slot') continue;
    const key = LEGACY_SLOT_ANNOTATION_COLUMNS[cellText(row.key)];
    if (key) {
      row.key = key;
    }
  }
}

function slotAnnotationColumns(row: Row) {
  return Object.keys(row)
    .filter((key) => isSlotAnnotationColumn(key))
    .map((key) => annotationColumnForKey(annotationKeyForColumn(key)));
}

function annotationColumnForKey(key: string) {
  const normalizedKey = LEGACY_SLOT_ANNOTATION_COLUMNS[key] ?? key;
  return normalizedKey ? `${SLOT_ANNOTATION_COLUMN_PREFIX}${normalizedKey}` : '';
}

function annotationKeyForColumn(column: string) {
  const key = column.startsWith(SLOT_ANNOTATION_COLUMN_PREFIX)
    ? column.slice(SLOT_ANNOTATION_COLUMN_PREFIX.length)
    : column.slice(LEGACY_SLOT_ANNOTATION_COLUMN_PREFIX.length);
  return LEGACY_SLOT_ANNOTATION_COLUMNS[key] ?? key;
}

function migrateLegacySlotAnnotationColumns(row: Row) {
  for (const column of Object.keys(row)) {
    if (!isLegacySlotAnnotationColumn(column)) continue;
    const canonicalColumn = annotationColumnForKey(annotationKeyForColumn(column));
    if (!cellText(row[canonicalColumn]) && cellText(row[column])) {
      row[canonicalColumn] = row[column];
    }
    delete row[column];
  }
}

function isSlotAnnotationColumn(column: string) {
  return (
    (column.startsWith(SLOT_ANNOTATION_COLUMN_PREFIX) && column !== SLOT_ANNOTATION_COLUMN_PREFIX) ||
    isLegacySlotAnnotationColumn(column)
  );
}

function isLegacySlotAnnotationColumn(column: string) {
  return (
    column.startsWith(LEGACY_SLOT_ANNOTATION_COLUMN_PREFIX) &&
    column !== LEGACY_SLOT_ANNOTATION_COLUMN_PREFIX
  );
}

function parseMapping(value: unknown): Record<string, Row[string]> {
  if (!value) return {};
  if (typeof value === 'object' && !Array.isArray(value)) {
    return value as Record<string, Row[string]>;
  }
  try {
    const parsed = JSON.parse(String(value));
    return parsed && typeof parsed === 'object' && !Array.isArray(parsed) ? parsed : {};
  } catch {
    return parseSimpleMapping(String(value));
  }
}

function parseSimpleMapping(value: string): Record<string, Row[string]> {
  const parsed: Record<string, Row[string]> = {};
  for (const line of value.split('\n')) {
    const separatorIndex = line.indexOf(':');
    if (separatorIndex <= 0) return {};
    const key = line.slice(0, separatorIndex).trim();
    const rawValue = line.slice(separatorIndex + 1).trim();
    if (!key) return {};
    parsed[key] = rawValue.replace(/^['"]|['"]$/g, '');
  }
  return parsed;
}

function splitList(value: unknown) {
  return cellText(value)
    .split(';')
    .map((part) => part.trim())
    .filter(Boolean);
}

function cloneTables(tables: Tables): Tables {
  return Object.fromEntries(
    Object.entries(tables).map(([name, rows]) => [name, rows.map((row) => ({ ...row }))])
  );
}

function cellText(value: unknown): string {
  return value === null || value === undefined ? '' : String(value).trim();
}

File: src/dataharmonizer_template_builder/table_sync.py
"""Synchronize redundant references across editable Schemasheets tables."""

from __future__ import annotations

import copy
import json
from collections.abc import Iterable, Mapping
from typing import Any

import yaml

from dataharmonizer_template_builder.models import Diagnostic, JsonDict, TableRows

CLASS_TABLE = "classes"
SLOT_TABLE = "slots"
ENUM_TABLE = "enums"
PERMISSIBLE_VALUE_TABLE = "permissible_values"
ANNOTATION_TABLE = "annotations"

SLOT_ANNOTATION_COLUMN_PREFIX = "Annotation: "
LEGACY_SLOT_ANNOTATION_COLUMN_PREFIX = "annotation_"

LEGACY_SLOT_ANNOTATION_COLUMNS = {
    "mimicc_default_unit": "default_unit",
}

LEGACY_SLOT_ROW_COLUMNS = {
    "annotation_mimicc_default_unit": "Annotation: default_unit",
}


def sync_tables(
    tables: TableRows,
    source_table: str | None = None,
) -> tuple[TableRows, list[Diagnostic]]:
    """Return tables with redundant references synchronized."""
    synced = copy.deepcopy(tables)
    synced.setdefault(CLASS_TABLE, [])
    synced.setdefault(SLOT_TABLE, [])
    synced.setdefault(ENUM_TABLE, [])
    synced.setdefault(PERMISSIBLE_VALUE_TABLE, [])
    synced.setdefault(ANNOTATION_TABLE, [])

    _sync_class_slots(synced)
    _sync_slot_annotations(synced, source_table)
    diagnostics = _sync_enum_annotations(synced, source_table)
    diagnostics.extend(_reference_diagnostics(synced))
    return synced, diagnostics


def _sync_class_slots(tables: TableRows) -> None:
    slots_by_class: dict[str, list[str]] = {}
    for row in tables[SLOT_TABLE]:
        class_name = _cell(row.get("class"))
        slot_name = _cell(row.get("slot"))
        if class_name and slot_name:
            slots_by_class.setdefault(class_name, []).append(slot_name)

    for row in tables[CLASS_TABLE]:
        class_name = _cell(row.get("class"))
        if class_name in slots_by_class:
            row["slots"] = "; ".join(slots_by_class[class_name])


def _sync_slot_annotations(tables: TableRows, source_table: str | None) -> None:
    annotation_rows = tables[ANNOTATION_TABLE]
    slot_rows_by_name = {
        _cell(row.get("slot")): row for row in tables[SLOT_TABLE] if _cell(row.get("slot"))
    }
    _migrate_legacy_slot_row_columns(slot_rows_by_name.values())
    _migrate_legacy_slot_annotations(annotation_rows)

    if source_table != ANNOTATION_TABLE:
        for slot_name, slot_row in slot_rows_by_name.items():
            for row_key in _slot_annotation_columns(slot_row):
                annotation_key = _annotation_key_for_column(row_key)
                value = _cell(slot_row.get(row_key))
                _upsert_annotation(annotation_rows, "slot", slot_name, annotation_key, value)
    else:
        for slot_row in slot_rows_by_name.values():
            for row_key in _slot_annotation_columns(slot_row):
                slot_row[row_key] = ""

    for row in annotation_rows:
        if _cell(row.get("element_type")) != "slot":
            continue
        slot_row = slot_rows_by_name.get(_cell(row.get("element")))
        row_key = _annotation_column_for_key(_cell(row.get("key")))
        if slot_row is not None and row_key:
            slot_row[row_key] = row.get("value", "")


def _sync_enum_annotations(tables: TableRows, source_table: str | None) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []
    annotation_rows = tables[ANNOTATION_TABLE]
    enum_rows_by_name = {
        _cell(row.get("enum")): row for row in tables[ENUM_TABLE] if _cell(row.get("enum"))
    }

    if source_table != ANNOTATION_TABLE:
        for row_index, (enum_name, enum_row) in enumerate(enum_rows_by_name.items(), start=1):
            parsed, ok = _parse_mapping_cell(enum_row.get("annotations"))
            if not ok:
                diagnostics.append(
                    Diagnostic(
                        "warning",
                        f"Enum {enum_name} has malformed annotations mapping.",
                        ENUM_TABLE,
                        row_index,
                        "annotations",
                    )
                )
                continue
            for key, value in parsed.items():
                _upsert_annotation(annotation_rows, "enum", enum_name, str(key), str(value))

    for row in annotation_rows:
        if _cell(row.get("element_type")) != "enum":
            continue
        enum_row = enum_rows_by_name.get(_cell(row.get("element")))
        key = _cell(row.get("key"))
        if enum_row is None or not key:
            continue
        parsed, _ = _parse_mapping_cell(enum_row.get("annotations"))
        parsed[key] = row.get("value", "")
        enum_row["annotations"] = json.dumps(parsed, ensure_ascii=False) if parsed else ""

    return diagnostics


def _reference_diagnostics(tables: TableRows) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []
    class_names = {
        _cell(row.get("class")) for row in tables[CLASS_TABLE] if _cell(row.get("class"))
    }
    slot_names = {_cell(row.get("slot")) for row in tables[SLOT_TABLE] if _cell(row.get("slot"))}
    enum_names = {_cell(row.get("enum")) for row in tables[ENUM_TABLE] if _cell(row.get("enum"))}

    for index, row in enumerate(tables[SLOT_TABLE], start=1):
        class_name = _cell(row.get("class"))
        if class_name and class_name not in class_names:
            diagnostics.append(
                Diagnostic(
                    "warning",
                    f"Slot references missing class {class_name}.",
                    SLOT_TABLE,
                    index,
                    "class",
                )
            )
        enum_name = _cell(row.get("range"))
        if enum_name.endswith("Menu") and enum_name not in enum_names:
            diagnostics.append(
                Diagnostic(
                    "warning",
                    f"Slot references missing enum {enum_name}.",
                    SLOT_TABLE,
                    index,
                    "range",
                )
            )

    for index, row in enumerate(tables[PERMISSIBLE_VALUE_TABLE], start=1):
        enum_name = _cell(row.get("enum"))
        if enum_name and enum_name not in enum_names:
            diagnostics.append(
                Diagnostic(
                    "warning",
                    f"Permissible value references missing enum {enum_name}.",
                    PERMISSIBLE_VALUE_TABLE,
                    index,
                    "enum",
                )
            )

    for index, row in enumerate(tables[ANNOTATION_TABLE], start=1):
        element_type = _cell(row.get("element_type"))
        element = _cell(row.get("element"))
        if not element_type or not element:
            continue
        if element_type == "class" and element not in class_names:
            diagnostics.append(
                Diagnostic("warning", "Annotation target was not found.", ANNOTATION_TABLE, index)
            )
        elif element_type == "slot" and element not in slot_names:
            diagnostics.append(
                Diagnostic("warning", "Annotation target was not found.", ANNOTATION_TABLE, index)
            )
        elif element_type == "enum" and element not in enum_names:
            diagnostics.append(
                Diagnostic("warning", "Annotation target was not found.", ANNOTATION_TABLE, index)
            )

    return diagnostics


def _upsert_annotation(
    rows: list[JsonDict],
    element_type: str,
    element: str,
    key: str,
    value: Any,
) -> None:
    match = next(
        (
            row
            for row in rows
            if _cell(row.get("element_type")) == element_type
            and _cell(row.get("element")) == element
            and _cell(row.get("key")) == key
        ),
        None,
    )
    if value in (None, ""):
        if match is not None:
            rows.remove(match)
        return
    if match is None:
        rows.append({"element_type": element_type, "element": element, "key": key, "value": value})
    else:
        match["value"] = value


def _migrate_legacy_slot_annotations(rows: list[JsonDict]) -> None:
    for row in rows:
        if _cell(row.get("element_type")) != "slot":
            continue
        key = LEGACY_SLOT_ANNOTATION_COLUMNS.get(_cell(row.get("key")))
        if key:
            row["key"] = key


def _migrate_legacy_slot_row_columns(rows: Iterable[JsonDict]) -> None:
    for row in rows:
        _migrate_legacy_slot_annotation_columns(row)
        for legacy_key, row_key in LEGACY_SLOT_ROW_COLUMNS.items():
            if not _cell(row.get(row_key)) and _cell(row.get(legacy_key)):
                row[row_key] = row[legacy_key]
            row.pop(legacy_key, None)


def _parse_mapping_cell(value: Any) -> tuple[dict[str, Any], bool]:
    if not value:
        return {}, True
    if isinstance(value, Mapping):
        return dict(value), True
    try:
        parsed = yaml.safe_load(str(value))
    except yaml.YAMLError:
        return {}, False
    return (dict(parsed), True) if isinstance(parsed, Mapping) else ({}, False)


def _cell(value: Any) -> str:
    return "" if value is None else str(value).strip()


def _slot_annotation_columns(row: Mapping[str, Any]) -> list[str]:
    return [
        _annotation_column_for_key(_annotation_key_for_column(key))
        for key in row
        if _is_slot_annotation_column(key)
    ]


def _annotation_column_for_key(key: str) -> str:
    normalized_key = LEGACY_SLOT_ANNOTATION_COLUMNS.get(key, key)
    return f"{SLOT_ANNOTATION_COLUMN_PREFIX}{normalized_key}" if normalized_key else ""


def _annotation_key_for_column(column: str) -> str:
    if column.startswith(SLOT_ANNOTATION_COLUMN_PREFIX):
        key = column.removeprefix(SLOT_ANNOTATION_COLUMN_PREFIX)
    else:
        key = column.removeprefix(LEGACY_SLOT_ANNOTATION_COLUMN_PREFIX)
    return LEGACY_SLOT_ANNOTATION_COLUMNS.get(key, key)


def _migrate_legacy_slot_annotation_columns(row: JsonDict) -> None:
    for column in list(row):
        if not _is_legacy_slot_annotation_column(column):
            continue
        canonical_column = _annotation_column_for_key(_annotation_key_for_column(column))
        if not _cell(row.get(canonical_column)) and _cell(row.get(column)):
            row[canonical_column] = row[column]
        row.pop(column, None)


def _is_slot_annotation_column(column: str) -> bool:
    return (
        column.startswith(SLOT_ANNOTATION_COLUMN_PREFIX)
        and column != SLOT_ANNOTATION_COLUMN_PREFIX
    ) or _is_legacy_slot_annotation_column(column)


def _is_legacy_slot_annotation_column(column: str) -> bool:
    return (
        column.startswith(LEGACY_SLOT_ANNOTATION_COLUMN_PREFIX)
        and column != LEGACY_SLOT_ANNOTATION_COLUMN_PREFIX
    )
